include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -1486,9 +1485,9 @@ __pattern_partial_sort_copy(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&&
     {
         // If our output buffer is larger than the input buffer, simply copy elements to the output and use
         // full sort on them.
-        auto __out_end = __pattern_walk2</*_IsSync=*/::std::false_type>(
-            __tag, __par_backend_hetero::make_wrapped_policy<__initial_copy_1>(__exec), __first, __last, __out_first,
-            __brick_copy<__hetero_tag<_BackendTag>, _ExecutionPolicy>{});
+        auto __out_end =
+            __pattern_walk2(__tag, __par_backend_hetero::make_wrapped_policy<__initial_copy_1>(__exec), __first, __last,
+                            __out_first, __brick_copy<__hetero_tag<_BackendTag>, _ExecutionPolicy>{});
 
         // Use regular sort as partial_sort isn't required to be stable
         __pattern_sort(
```

Comment:
So, since we are not wanting to do some sort of large refactor to use event dependencies to pipeline patterns at this point...

What about something like this to avoid the perf degradation:
Create something like `__type_requires_exterior_synchronization<__out_first>` which would resolve to `true_type` for anything for non-sycl backend, and for sycl backend it could return `bool_constant<!is_sycl_iterator_v<__out_first>>`.

We can do this without mentioning sycl at this level.

Reply:
This still leaves a lot to be desired vs the refactor (for host_iterators especially), but it at least removes the degradation.

Reply:
`is_sycl_iterator_v`...
What if we have a "fancy pipe" over a sycl_iterator? or what should we do in case of mix? zip(zip(sycl_iter, usm_ptr), usm_ptr) and so on?

Reply:
Yes, it would miss these, its not a perfect solution.  It provides some improvement though for raw sycl_iterator or legacy hetero iterators. The better solution is using events which can be passed as dependencies to patterns and then into sycl kernels.  

The solution described could be made better by adding a `__contains_sycl_iterator` which has specializations for our fancy iterators to pass through and check if any iterator child `__contains_sycl_iterator` and would therefore provide the synchronization.
zip and permutation would return the OR of their children iterators, transform would return its base's result.  If it is not a fancy iterator, it would just be defined as described above.

Reply:
Actually its a bit more complicated because we need the access pattern to actually dictate a dependency.  Probably permutation iterator should only pass-through to its base, not its map, because the map will be read-only.

It is of course the user of this structure's responsibility to make sure the usage of the sequence will generate the synchronization, but permutation iterator wont require write access for it's map even if it is used in write mode.

Read-only followed by read-only wouldn't require an ordering.  We need to be most careful to avoid believing there is a synchronization where there isn't one.